openedx/openedx-platform
# third_party_auth SAML/Shibboleth Provider Support

**Description**
This is the first part of implementing the new Shibboleth SSO support integrated into third_party_auth (TPA). See the [technical spec](https://docs.google.com/document/d/1YX_DFfDrU1JxkSME3ha1YpX9mc9AKO5MUMjfLXqmf38/edit) for a lot more background and discussion.

This PR will merge the new code to a feature branch, `feature/shibboleth-tpa`

**Changes in this PR**
This first PR does the following:
- Upgrades `third_party_auth` (TPA) to use the latest version of python-social-auth (this was a non-trivial upgrade)
- Adds a new TPA provider class that can be used as a base class for writing SAML providers
- Includes modified versions of python-saml and python-social-auth that do the heavy lifting of implementing SAML support. Changes are being contributed to those upstream projects.
- Includes two example providers, both using [TestShib](https://www.testshib.org/), the public Shibboleth test server.
- Modifies the TPA code so that multiple providers can use the same python-social-auth "backend" class.

**New Requirement**
This introduces one new requirement: [python-saml](https://github.com/onelogin/python-saml), which also [currently] requires two new debian packages not installed on most edX instances: `libxmlsec1-dev` and `swig`.

**Dependencies**
- [x] https://github.com/edx/configuration/pull/2002
- [x] https://github.com/onelogin/python-saml/pull/66 (external upstream PR)
- [ ] https://github.com/onelogin/python-saml/pull/67 (external upstream PR)
- [ ] https://github.com/onelogin/python-saml/pull/68 (external upstream PR)
- [ ] https://github.com/onelogin/python-saml/pull/69 (external upstream PR)
- [ ] https://github.com/omab/python-social-auth/pull/616 (external upstream PR)

**Sandbox**
- http://sandbox5.opencraft.com/login
- http://sandbox5.opencraft.com/auth/saml/metadata.xml

**Internal Code Review PR**: https://github.com/edx/edx-platform/pull/7893

**Upgrade Issue**
When merged, this PR does cause a problem for existing installations: `python-social-auth` at one point converted from `syncdb` to south/django migrations, and bumping the version of `python-social-auth` that we use will result in this error next time `paver update_db` is run:

```
django.db.utils.DatabaseError: table "social_auth_usersocialauth" already exists
```

To fix this (on devstack), the user must manually run this command:

```
./manage.py lms migrate default 0001_initial --fake --settings=devstack
```

Unfortunately, I'm not aware of any way to avoid this issue other than this manual fix. New installs are not affected.

**Testing Instructions**:
Go to the sandbox link above and login via TestShib. Try registering then linking and unlinking TestShib accounts. If it doesn't work, it may be necessary to re-upload the metadata to TestShib's "Register" page. A link to the metadata is above.

To test locally, see https://github.com/edx/edx-platform/pull/7893

**Not included**:
The following will be coming in future PRs against this feature branch:
- provider configuration will be moved to the database
- Metadata fetching
- Fix for [CRI-9](https://openedx.atlassian.net/browse/CRI-9)
- Ability to specify default third_party_auth provider via query param
- UX changes: Allow to configure multiple Shibboleth identity providers as "secondary" providers
